src/lib/core/datetime.c

```diff
@@ -142,3 +143,154 @@ datetime_compare(const struct datetime *lhs, const struct datetime *rhs)
 
 	return COMPARE_RESULT(lhs->nsec, rhs->nsec);
 }
+
+/**
+ * Interval support functions: stringization and operations
+ */
+
+#define NANOS_PER_SEC 1000000000
+
+static inline int64_t
+interval_seconds(const struct datetime_interval *ival)
+{
+       return (int64_t)ival->sec + ival->nsec / NANOS_PER_SEC;
+}
+
+static inline int
+interval_days(const struct datetime_interval *ival)
+{
+       return interval_seconds(ival) / SECS_PER_DAY;
+}
+
+static inline int
+interval_hours(const struct datetime_interval *ival)
+{
+       return interval_seconds(ival) / (60 * 60);
+}
+
+static inline int
+interval_minutes(const struct datetime_interval *ival)
+{
+       return interval_seconds(ival) / 60;
+}
+
+/**
+ * if |nsec| is larger than allowed range 1e9 then modify passed
+ * sec accordingly
+ */
+static void
+denormalize_interval_nsec(int64_t *psec, int *pnsec)
+{
+	assert(psec != NULL);
+	assert(pnsec != NULL);
+	int64_t sec = *psec;
+	int nsec = *pnsec;
+	/*
+	 * nothing to change:
+	 * - if there is small nsec with 0 in sec
+	 * - or if both sec and nsec have the same sign, and nsec is
+	 *   small enough
+	 */
+	if (nsec == 0)
+		return;
+
+	bool zero_or_same_sign = sec == 0 || (sec < 0) == (nsec < 0);
+	if (zero_or_same_sign && labs(nsec) < NANOS_PER_SEC)
+		return;
+
+	int64_t normalized_nsec = sec * NANOS_PER_SEC + nsec;
+	sec = normalized_nsec / NANOS_PER_SEC;
+	nsec = normalized_nsec % NANOS_PER_SEC;
+
+	*psec = sec;
+	*pnsec = nsec;
+}
+
+/**
+ * SNPRINT-like function to generate textual representation of a sec.nsec
+ * given buffer and size of buffer
+ */
+static size_t
+seconds_str(char *buf, ssize_t len, bool need_sign, long sec, int nsec)
+{
+	sec %= 60;
+	denormalize_interval_nsec(&sec, &nsec);
+	size_t sz = 0;
+	bool is_neg = sec < 0 || (sec == 0 && nsec < 0);
+
+	SNPRINT(sz, snprintf, buf, len, is_neg ? "-" : (need_sign ? "+" : ""));
+	sec = labs(sec);
+	if (nsec != 0) {
+		SNPRINT(sz, snprintf, buf, len, "%" PRId64 ".%09ld seconds",
+			sec, labs(nsec));
+	} else {
+		SNPRINT(sz, snprintf, buf, len, "%" PRId64 " seconds", sec);
+	}
+	return sz;
+}
+
+#define SPACE() \
+	do { \
+		if (sz > 0) { \
+			SNPRINT(sz, snprintf, buf, len, ", "); \
+		} \
+	} while (0)
+
+size_t
+interval_to_string(const struct datetime_interval *ival, char *buf, ssize_t len)
```

Comment:
Let me clarify a bit how `SNPRINT` macro (which is used widely across a core) is working here:
- it essentially operating as usual `snprint` wrapper and is handling cases when `buf` and `size` were zero, in this case it simply evaluate the necessary size of buffer (and this return value might be used later with correspondingly [re]allocated buffer size);
- if there is non-zero `buf` and characters to print are still fitting in the size then `snprint` (or other function we used here) wil write to the buffer and remember location from which continue to write;
- but... otherwise (if there is NULL passed as `buf` or if output result is larger than passed buffer size) then no `buf` modified, and we only continue to count the necessary size for return value.

Please see how we use 2 pas approach for allocating target buffer memory, if result is not fitting into the default buffer size:
```lua
local function interval_tostring(self)
    check_interval(self, 'datetime.interval.tostring')
    local buff = ival_tostr_stash_take()
    local len = builtin.tnt_interval_to_string(self, buff, IVAL_TOSTRING_BUFSIZE)
    if len < IVAL_TOSTRING_BUFSIZE then
        local s = ffi.string(buff)
        ival_tostr_stash_put(buff)
        return s
    end
    -- slow path - reallocate for a fuller size, and then restart interval_to_string
    ival_tostr_stash_put(buff)
    buff = ffi.new('char[?]', len + 1)
    builtin.tnt_interval_to_string(self, buff, len + 1)
    return ffi.string(buff)
```

And returning to your question - this careful logics is not implementable without much manual operations across `snprint` calls. (And FWIW your version will be segfaulting if NULL will be passed as buffer, thus some tests will fail).